MathlibTest/Junk.lean

```diff
@@ -0,0 +1,360 @@
+and so is equivalent as far as theorem proving goes; but (if run in a program) causes a panic.
+
+This is currently deprecated in favor of `[i]!`, the `getElem!` syntax.
+
+Note that similar APIs are duplicated across several types: e.g. `Vector.get`, `Array.getD`,
+`String.get!`, `Option.getD`, etc.
+-/
+#guard_msgs(drop warning) in
+example : [100,200,300,400,500].get! 12 = default := by
+  rfl
+
+/- Trying to evaluate the expression leads to a panic. -/
+#guard_msgs(drop all) in
+#eval [100,200,300,400,500].get! 12
+
+/-! # Strings -/
+
+/-- `String.get`, `String.get!`, and `String.get?` generally mimic arrays and lists by accessing
+individual characters within the String, as if it were a `List Char`. The junk value is the
+character `'A'`.
+
+But there's the additional complication that UTF-8 characters (the encoding of Strings in Lean) are
+a variable number of bytes, so even though the index is a byte offset, this can point to an invalid
+location if it lands in the middle of a character.
+
+`String.get'` splits the difference by requiring a proof (or as the documentation calls it, merely
+"evidence") that the index is within bounds, but it will still give the junk `'A'` if it lands in
+the middle of a character. -/
+example :
+    "Hello!".get 0 = 'H' ∧
+    "Hello!".get (.mk 100) = default ∧
+    "Hello!".get! (.mk 100) = default ∧
+    "Hello!".get? (.mk 100) = none ∧
+    "Hello!".get' (.mk 3) (by decide) = 'l' ∧
+    "⋃ℕℐℂ₀𝔻ℰ".get' (.mk 2) (by decide) = default ∧
+    "⋃ℕℐℂ₀𝔻ℰ".get' (.mk 3) (by decide) = 'ℕ' ∧
+    default = 'A'
+    := by
+  and_intros <;> rfl
+
+/-! # Derivatives -/
+
+/-- The derivative of a function at a point where it's not differentiable is given the junk
+value of 0. In this example, the derivative of `7x + |x|` at `x = 0` is equal to 0 (even though
+the "sensible" values would be 6, 8, or maybe 7).
+
+This holds for the many variants of derivatives, like `fderiv`, `fderivWithin`, etc.
+
+Note this junk value can mean expressions like "twice differentiable" can be harder to express
+than you think: the function `if (x = 0) then 1 else x ^ 2` isn't differentiable at 0, but its
+derivative simplifies to `if (x = 0) then 0 else 2 * x`, which is just `2 * x`, and so its
+derivative is differentiable everywhere.
+
+In that example, instead of writing `Differentiable (deriv f)`, use `ContDiff ℝ 2 f`.
+-/
+example : deriv (fun (x : ℝ) ↦ 7 * x + |x|) 0 = 0 := by
+  apply deriv_zero_of_not_differentiableAt
+  by_contra h
+  have h₂ : DifferentiableAt ℝ (fun (x : ℝ) ↦ (-7) * x) 0 := by
+    fun_prop
+  apply not_differentiableAt_abs_zero
+  simpa [Pi.add_def] using h₂.add h
+
+/-- Another version is that `analyticOrderAt` and `analyticOrderNatAt` give the junk value
+0 when the function isn't analytic at the point. -/
+example : analyticOrderAt Real.log 0 = 0 := by
+  apply analyticOrderAt_of_not_analyticAt
+  by_contra h
+  replace h := AnalyticAt.continuousAt h
+  simp at h
+
```

Comment:
maybe an example of how this helps in proofs would be nice, like the canonical example of Rolle's theorem